File: matcomgrader/icpc_caribbean_qualifiers_2025/real_contest/m.py
"""
https://matcomgrader.com/problem/9830/matrix-operations/
"""

# Accepted: FIRST EVER SOLVE in the ICPC Caribbean 2025 Qualifiers

def main():
    rows, cols, q = [int(s) for s in input().split()]

    matrix = [[int(s) for s in input().split()] for _ in range(rows)]

    for _ in range(q):
        fields = [int(s) for s in input().split()]
        test_type = fields[0]

        if test_type == 1:
            r = fields[1] - 1
            matrix[r] = matrix[r][::-1]

        elif test_type == 2:
            c = fields[1] - 1

            for r in range(rows // 2):
                matrix[r][c], matrix[rows - r - 1][c] = matrix[rows - r - 1][c], matrix[r][c]

        else:
            r, c = fields[1:]
            print(matrix[r - 1][c - 1])


main()

File: matcomgrader/icpc_caribbean_qualifiers_2025/real_contest/m.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int rows, cols, queries;
    cin >> rows >> cols >> queries;

    vector<vector<int>> matrix(rows);

    for (int r = 0; r < rows; ++r) {
        vector<int> row(cols);
        for (int c = 0; c < cols; ++c) {
            cin >> row[c];
        }
        matrix[r] = row;
    }

    for (int _q = 0; _q < queries; ++_q) {
        int operation;
        cin >> operation;

        if (operation == 1) {
            int r;
            cin >> r;
            r -= 1;

            for (int i = 0; i < cols / 2; ++i) {
                int temp = matrix[r][i];
                matrix[r][i] = matrix[r][cols - i - 1];
                matrix[r][cols - i - 1] = temp;
            }
        }
        else if (operation == 2) {
            int c;
            cin >> c;
            c -= 1;

            for (int i = 0; i < rows / 2; ++i) {
                int temp = matrix[i][c];
                matrix[i][c] = matrix[rows - i - 1][c];
                matrix[rows - i - 1][c] = temp;
            }
        }
        else {
            int r, c;
            cin >> r >> c;
            r -= 1;
            c -= 1;

            cout << matrix[r][c] << endl;
        }
    }
}